test/e2e/network/netpol/network_policy_api.go

```diff
@@ -209,4 +210,76 @@ var _ = common.SIGDescribe("Netpol API", func() {
 		framework.ExpectNoError(err)
 		framework.ExpectEqual(len(nps.Items), 0, "filtered list should have 0 items")
 	})
+
+	/*
+		Release: v1.21
+		Testname: NetworkPolicy support EndPort Field
+		Description:
+		- EndPort field cannot be defined if the Port field is not defined
+		- EndPort field cannot be defined if the Port field is defined as a named (string) port.
+		- EndPort field must be equal or greater than port.
+	*/
+	ginkgo.It("should support creating NetworkPolicy API with endport field [Feature:NetworkPolicyEndPort]", func() {
+
+		ns := f.Namespace.Name
+		npClient := f.ClientSet.NetworkingV1().NetworkPolicies(ns)
+
+		npTemplate := &networkingv1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "e2e-example-netpol-endport-validate",
+				Labels: map[string]string{
+					"special-label": f.UniqueName,
+				},
+			},
+			Spec: networkingv1.NetworkPolicySpec{
+				PodSelector: metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"pod-name": "test-pod",
+					},
+				},
+				PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeEgress},
+			},
+		}
+
+		ginkgo.By("EndPort field cannot be defined if the Port field is not defined.")
+		var endport int32 = 20000
+		egressRule := networkingv1.NetworkPolicyEgressRule{}
+		egressRule.Ports = append(egressRule.Ports, networkingv1.NetworkPolicyPort{EndPort: &endport})
+		npTemplate.Spec.Egress = append(npTemplate.Spec.Egress, egressRule)
+		_, err := npClient.Create(context.TODO(), npTemplate, metav1.CreateOptions{})
+		framework.ExpectError(err, "request template:%v", npTemplate)
```

Comment:
Interestingly this error is returning nil, you should be hitting here instead:

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/networking/validation/validation.go#L89

Reply:
real ? I have test this case using local-cluster-up. And here it pass 😂 . It should get error.

Reply:
I add log as fellow:

>INFO: template: &NetworkPolicy{ObjectMeta:{ e2e-example-netpol-endport-validate     0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[special-label:netpol-3563] map[] [] []  []},Spec:NetworkPolicySpec{PodSelector:{map[pod-name:test-pod] []},Ingress:[]NetworkPolicyIngressRule{},Egress:[]NetworkPolicyEgressRule{NetworkPolicyEgressRule{Ports:[]NetworkPolicyPort{NetworkPolicyPort{Protocol:nil,Port:<nil>,EndPort:*20000,},},To:[]NetworkPolicyPeer{},},},PolicyTypes:[Egress],},} err:NetworkPolicy.extensions "e2e-example-netpol-endport-validatetwg84" is invalid: spec.egress[0].ports[0].endPort: Invalid value: 20000: may not be specified when `port` is not specified

Reply:
Is it that right? 😂 
@knabben

Reply:
Yes it should take a look in the failing CI job

Reply:
maybe you need to enable the feature gate :)

Reply:
I add the describe mark label "[Feature:NetworkPolicyEndPort]". So seems that it still will be run. 
Which place can we add to open  the feature gate before running it ?
@rikatz @knabben

Reply:
add a feature gate check like 

https://github.com/kubernetes/kubernetes/blob/4740173f3378ef9d0dc59b0aa9299444a97d0818/test/e2e/framework/skipper/skipper.go#L139-L143

thanks for inspiring from @aojea